sonar-plugin-api/src/main/java/org/sonar/api/measures/MetricDefinition.java

```diff
@@ -0,0 +1,265 @@
+
+    private void registerMetric(NewMetricImpl newMetric) {
+      // TODO check metric does not already exists
+      metricsByKey.put(newMetric.key, new MetricImpl(newMetric));
+    }
+  }
+
+  interface NewMetric {
+    NewMetric setKey(String s);
+
+    NewMetric setName(String s);
+
+    // TODO add all fields needed on metric : domain, type, best value, direction, qualitative, etc.
+
+    NewComputer createMeasureComputer();
+
+    void done();
+  }
+
+  class NewMetricImpl implements NewMetric {
+    private final NewMetricContext context;
+
+    private String key;
+
+    private String name;
+
+    private Computer computer;
+
+    public NewMetricImpl(NewMetricContext context) {
+      this.context = context;
+    }
+
+    @Override
+    public NewMetric setKey(String s) {
+      this.key = s;
+      return this;
+    }
+
+    @Override
+    public NewMetric setName(String s) {
+      this.name = s;
+      return this;
+    }
+
+    @Override
+    public NewComputer createMeasureComputer() {
+      return new NewComputerImpl(this);
+    }
+
+    void setNewComputerImpl(NewComputerImpl newComputer){
+      this.computer = new ComputerImpl(newComputer);
+    }
+
+    @Override
+    public void done() {
+      // TODO fail if no computer
+      context.registerMetric(this);
+    }
+  }
+
+  interface Metric {
+    String getKey();
+
+    String getName();
+
+    Computer getComputer();
+  }
+
+  class MetricImpl implements Metric {
+    private final String key;
+    private final String name;
+    private final Computer computer;
+
+    public MetricImpl(NewMetricImpl newMetric) {
+      this.key = newMetric.key;
+      this.name = newMetric.name;
+      this.computer = newMetric.computer;
+    }
+
+    @Override
+    public String getKey() {
+      return key;
+    }
+
+    @Override
+    public String getName() {
+      return name;
+    }
+
+    @Override
+    public Computer getComputer() {
+      return computer;
+    }
+
+    @Override
+    public String toString() {
+      return "Metric{" +
+        "key='" + key + '\'' +
+        '}';
+    }
+  }
+
+  interface NewComputer {
+    NewComputer setInputMetricKeys(String... metricKeys);
+
+    NewComputer setMeasureComputer(MeasureComputer measureComputer);
+
+    NewMetric done();
+  }
+
+  class NewComputerImpl implements NewComputer {
+
+    private final NewMetricImpl newMetric;
+
+    public NewComputerImpl(NewMetricImpl newMetric) {
+      this.newMetric = newMetric;
+    }
+
+    private String[] inputMetricKeys;
+    private MeasureComputer measureComputer;
+
+    @Override
+    public NewComputer setInputMetricKeys(String... metricKeys) {
+      this.inputMetricKeys = metricKeys;
+      return this;
+    }
+
+    @Override
+    public NewComputer setMeasureComputer(MeasureComputer measureComputer) {
+      this.measureComputer = measureComputer;
+      return this;
+    }
+
+    @Override
+    public NewMetric done() {
+      // TODO fail if no metrics or no computer
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)